### PR TITLE
fix: add port retry mechanism for server startup with logging and failure handling

### PR DESCRIPTION
**Pull request by Vidhushaaa30**
## Description
This PR improves the reliability of server startup in `run_server.py` by introducing a retry mechanism for port binding.

Instead of attempting to bind to a single port once, the server now retries binding on subsequent ports if the initial port is unavailable. Logging has been added for each retry attempt, and a clear error is raised if all attempts fail.

## Motivation and Context
Previously, if the specified port was already in use, the server would fail immediately without attempting recovery. This could interrupt workflows when restarting the server or running multiple services locally.

This change improves usability and robustness by:
- Retrying port binding on subsequent ports
- Providing logging for better visibility into retry attempts
- Ensuring a clear failure message when all attempts are exhausted
- Preserving existing behavior for both local (`bind_local`) and default modes

## How Has This Been Tested?
- Started the server on an available port to confirm normal behavior remains unchanged
- Simulated port conflicts by running another process on the default port and verified that:
  - The server retries with the next available ports
  - Warning logs are printed for each retry attempt
  - The server successfully starts on an available port
- Verified that when all retry attempts are exhausted, a RuntimeError is raised with a clear message
- Ensured no regression in functionality by running the server before and after changes and comparing behavior

## Screenshots (if appropriate):
N/A

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Code refactor or cleanup (changes to existing code for improved readability or performance)

## Checklist:
- [ ] I adapted the version number under `py/visdom/VERSION` according to Semantic Versioning
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

## Summary by Sourcery

Improve robustness of server startup by adding a retry mechanism for port binding with clear failure reporting.

Bug Fixes:
- Handle server startup failures when the requested port is already in use by retrying on subsequent ports instead of failing immediately.

Enhancements:
- Add configurable retry loop for binding to incremental ports, with warning logs for each failed attempt and a final runtime error if all attempts are exhausted.

### CODE REVIEW

**Comment by rajnisht7** [comment]
have you tried the random approach instead of continuous port incremental approach

File: py/visdom/server/run_server.py [hunk 1]
```diff
@@ -49,10 +49,29 @@ def start_server(
...
-        app.listen(port, max_buffer_size=1024**3)
+    initial_port = port
+    max_attempts = 10
+    attempt = 0
+    bound = False
+
+    while attempt < max_attempts:
+      try:
+        if bind_local:
+            app.listen(port, max_buffer_size=1024**3, address="127.0.0.1")
```
Comment (rajnisht7) [inline py/visdom/server/run_server.py]: isnt buffer size of 1GB is quite large

**Comment by rajnisht7** [comment]
@Vidhushaaa30 kindly address copilot reviews

File: py/visdom/server/run_server.py [hunk 2]
```diff
@@ -49,10 +49,29 @@ def start_server(
...
-    if bind_local:
-        app.listen(port, max_buffer_size=1024**3, address="127.0.0.1")
-    else:
-        app.listen(port, max_buffer_size=1024**3)
+    initial_port = port
+    max_attempts = 10
+    attempt = 0
+    bound = False
+
+    while attempt < max_attempts:
```
Comment (Manik-Khajuria-5) [inline py/visdom/server/run_server.py:57]: @Vidhushaaa30 Can u please fix the identation here

File: py/visdom/server/run_server.py [hunk 3]
```diff
@@ -49,10 +49,29 @@ def start_server(
         use_frontend_client_polling=use_frontend_client_polling,
         eager_data_loading=eager_data_loading,
     )
-    if bind_local:
-        app.listen(port, max_buffer_size=1024**3, address="127.0.0.1")
-    else:
-        app.listen(port, max_buffer_size=1024**3)
+    initial_port = port
+    max_attempts = 10
+    attempt = 0
```
[the remark this answers is not in the document]
Comment (Manik-Khajuria-5) [reply py/visdom/server/run_server.py:54]: @Vidhushaaa30 Max number of port must not be hardcoded

**Review by Manik-Khajuria-5 — CHANGES_REQUESTED** [review]
@Vidhushaaa30 Can u please handle edge cases properly and add proper checking for Type error , value error and other type of errors

**Comment by mariobehling** [comment]
Thanks for the contribution.

We are closing this PR because the current implementation is not ready to be merged and the requested review changes have not been addressed.

The idea of improving server startup behavior when a port is already in use can be useful. However, the current PR still has several issues:

* The retry behavior is hardcoded with `max_attempts = 10`
* The implementation retries ports incrementally without clearly justifying this approach
* The final bound port handling and logging need to be clearer
* Edge cases such as invalid port values and related error handling are not sufficiently covered
* There were also indentation and code quality concerns raised during review
* The PR does not yet provide a sufficiently robust and maintainable implementation for server startup logic

For this kind of change, we need a carefully designed solution because automatic port switching changes runtime behavior. Users and scripts may expect the server to bind to the requested port or fail clearly. If Visdom silently starts on another port, the client side and logs must make this explicit and reliable.

A new PR would be welcome if it:

* Defines the expected behavior clearly
* Makes retry behavior configurable instead of hardcoded
* Handles invalid inputs and relevant edge cases properly
* Logs the final selected port clearly
* Ensures client/server behavior remains consistent
* Includes tests or a clear reproducible verification case

Closing this PR for now. Thanks again for working on it.
